### PR TITLE
Generate Systemd

### PR DESCRIPTION
**Pull request by baude**
Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/systemdgen/systemdgen.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+var template string = `[Unit]
+Description=%s Podman Container
+[Service]
+Restart=%s
+ExecStart=/usr/bin/podman start %s
+ExecStop=/usr/bin/podman stop %s
+KillMode=none
+Type=forking
+PIDFile=%s
```
Comment (EmilienM) [inline pkg/systemdgen/systemdgen.go]: I've tested this config and I confirm it is what works the best for us in OpenStack.

File: pkg/systemdgen/systemdgen.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/pkg/errors"
+	"path/filepath"
+)
+
+var template string = `[Unit]
+Description=%s Podman Container
+[Service]
+Restart=%s
+ExecStart=/usr/bin/podman start %s
+ExecStop=/usr/bin/podman stop %s
```
Comment (EmilienM) [inline pkg/systemdgen/systemdgen.go]: we usually add "-t 10" to all our containers to give the container some time to stop properly.

Comment (mheon) [reply pkg/systemdgen/systemdgen.go]: Maybe make timeout configurable? Podman generate systemd --timeout <x>?

Comment (EmilienM) [reply pkg/systemdgen/systemdgen.go]: > Maybe make timeout configurable? Podman generate systemd --timeout ?

--stop-timeout then, if we have an option

File: cmd/podman/generate_systemd.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+`,
+	}
+)
+
+func init() {
+	containerSystemdCommand.Command = _containerSystemdCommand
+	containerSystemdCommand.SetHelpTemplate(HelpTemplate())
+	containerSystemdCommand.SetUsageTemplate(UsageTemplate())
+	flags := containerSystemdCommand.Flags()
+	//flags.BoolVarP(&containerKubeCommand.Service, "service", "s", false, "Generate YAML for kubernetes service object")
```
Comment (rhatdan) [inline cmd/podman/generate_systemd.go]: Remove this comment.

File: pkg/adapter/containers.go [hunk 4]
```diff
@@ -607,3 +608,15 @@ func (r *LocalRuntime) Start(ctx context.Context, c *cliconfig.StartValues, sigP
...
+func (r *LocalRuntime) GenerateSystemd(c *cliconfig.GenerateSystemdValues) (string, error) {
+	ctr, err := r.Runtime.LookupContainer(c.InputArgs[0])
+	if err != nil {
+		return "", err
+	}
+	name := ctr.ID()
+	if c.Name {
+		name = ctr.Name()
+	}
+	return systemdgen.CreateSystemdUnitAsString(name, "always", ctr.Config().StaticDir)
```
Comment (rhatdan) [inline pkg/adapter/containers.go]: I think the default should be on-failure.

File: pkg/systemdgen/systemdgen.go (hunk 2 above)
Comment (rhatdan) [reply pkg/systemdgen/systemdgen.go]: I like that.
I think we should also have --restart-policy and default to on-failure.

**Comment by mheon** [comment]
@baude Can you hit the `podman run` manpage, remove the default unit file in the examples, and recommend they use this instead?

File: cmd/podman/generate_systemd.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"fmt"
+	"github.com/containers/libpod/cmd/podman/cliconfig"
```
Comment (mheon) [inline cmd/podman/generate_systemd.go]: Newline between these two

**Comment by baude** [comment]
I just put this PR up so I could get feedback from @EmilienM ... still lots of things to do.

File: cmd/podman/generate_systemd.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/spf13/cobra"
+)
+
+var (
+	containerSystemdCommand     cliconfig.GenerateSystemdValues
+	containerSystemdDescription = `Command generates a systemd unit file for a Podman container
+  `
+	_containerSystemdCommand = &cobra.Command{
+		Use:   "systemd [flags] CONTAINER | POD",
+		Short: "Generate a systemd until file for a Podman container",
```
Comment (adrianreber) [inline cmd/podman/generate_systemd.go]: Replace 'until file' with 'unit file'

**Comment by baude** [comment]
updated and removed the WIP

File: docs/podman-generate-systemd.1.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+$ sudo podman generate systemd --restart-policy=always -t 1 nginx
+[Unit]
+Description=c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc Podman Container
+[Service]
+Restart=always
+ExecStart=/usr/bin/podman start c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
+ExecStop=/usr/bin/podman stop -t 1 c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
+KillMode=none
+Type=forking
+PIDFile=/var/lib/containers/storage/overlay-containers/c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc/userdata/c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc.pid
```
Comment (jamescassell) [inline docs/podman-generate-systemd.1.md]: Why the change in approach from the simpler example given in podman-run below?  i.e., why a pid file and why not the `-a` option to `podman start`?

**Comment by mheon** [comment]
More reliable way of tracking the actual container process - we've had
people report leaking processes when not using PID files

**Comment by rhatdan** [comment]
I think it would be best if this was template based, that way users could specify a different template that they want to use.

**Comment by baude** [comment]
> I think it would be best if this was template based, that way users could specify a different template that they want to use.

I dont know what you mean .... please elaborate.

**Comment by rhatdan** [comment]
You hard coded a specific unit file format into the tool.  I just want you to take that format, make it a template and then allow me to override it.

Then if I want to setup different types of containers running in Unit files, I could do it.

File: cmd/podman/generate_systemd.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+	// User input stop timeout must be 0 or greater
+	if c.Flag("timeout").Changed && c.StopTimeout < 0 {
+		return errors.New("timeout value must be 0 or greater")
+	}
+	// Make sure the input restart policy is valid
+	if err := systemdgen.ValidateRestartPolicy(c.RestartPolicy); err != nil {
+		return err
+	}
+
```
Comment (edsantiago) [inline cmd/podman/generate_systemd.go]: Please add minimal arg checking.
```
$ ./bin/podman generate systemd
panic: runtime error: index out of range
...stack trace
```

File: pkg/adapter/containers.go [hunk 9]
```diff
@@ -5,6 +5,7 @@ package adapter
 import (
 	"context"
 	"fmt"
+	"github.com/containers/libpod/pkg/systemdgen"
```
Comment (mheon) [inline pkg/adapter/containers.go]: Ordering

**Comment by mheon** [comment]
Two comments and a rebase and this ought to be good

**Comment by mheon** [comment]
@baude Mind looking at this one? Needs a bit more work to make it in

File: cmd/podman/generate_systemd.go (hunk 8 above)
Comment (baude) [reply cmd/podman/generate_systemd.go]: thanks ed!. got it

File: docs/podman-generate-systemd.1.md (hunk 7 above)
Comment (baude) [reply docs/podman-generate-systemd.1.md]: it all has to do with reliability and control of figuring out if the container dies.
